### PR TITLE
Validate Genesis Root

### PR DESCRIPTION
**Pull request by jacque006**
Resolves https://github.com/thehubbleproject/hubble-contracts/issues/559

### TODO
- [x] Figure out why genesis state root validation is failing in client integration test.
- [x] Custom error types. ~and tests for failures~ Complexity and time of tests current not worth coverage of error cases. Likely worth revisiting in the future if we are able to stub out or isolate certain contracts and systems.

### CODE REVIEW

File: ts/client/coreAPI.ts [hunk 1]
```diff
@@ -73,7 +73,7 @@ export class CoreAPI implements ICoreAPI {
         const contracts = genesis.getContracts(signer);
         const syncedPoint = new SyncedPoint(
             genesis.auxiliary.genesisEth1Block,
-            0
+            -1
```
Comment (jacque006) [inline ts/client/coreAPI.ts:76]: Setting this to an uninitialized state and letting the genesis event sync move it to 0 seemed correct to me.

File: ts/client/services/events/genesisBatch.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        // Create L2 state
+        const genCommit = new GenesisCommitment(this.genesisStateRoot);
+        const genesisBatch = new ConcreteBatch([genCommit]);
+
+        // Verify match
+        if (l1GenesisBatch.commitmentRoot !== genesisBatch.commitmentRoot) {
+            throw new GenesisBatchCommitmentRootMismatch(
+                l1GenesisBatch.commitmentRoot,
+                genesisBatch.commitmentRoot
+            );
```
Comment (jacque006) [inline ts/client/services/events/genesisBatch.ts:100]: This and the validations/errors above are all the edge cases I could think of to verify the genesis batch, let me know if you know of other cases/things we should verify.

File: ts/client/services/events/genesisBatch.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+export class GenesisBatchEventSyncer extends ContractEventSyncer {
+    private readonly genesisStateRoot: string;
+    private readonly rollup: Rollup;
+    private readonly syncpoint: SyncedPoint;
+    private readonly batchStorage: BatchStorage;
+
+    constructor(api: CoreAPI) {
+        super(
+            api.contracts.rollup,
+            api.contracts.rollup.filters.NewBatch(null, null, null)
```
Comment (jacque006) [inline ts/client/services/events/genesisBatch.ts:30]: @ChihChengLiang Would have loved it if the 3rd filter param `batchType` was indexed, would makes this syncer bring in almost no other events. I assume it is not indexed to save gas?

File: ts/client/features/genesis.ts [hunk 4]
```diff
@@ -17,14 +22,49 @@ export class GenesisCommitment extends BaseCommitment {
...
+
+        // Create L2 state
         const commitment = new GenesisCommitment(this.genesisStateRoot);
-        return new ConcreteBatch([commitment]);
+        const batch = new ConcreteBatch([commitment]);
+
+        // Get L1 state
+        const l1Txn = await event.getTransaction();
+        const l1Batch = await this.rollup.getBatch(batchIDBN, {
+            blockTag: l1Txn.blockNumber
```
Comment (ChihChengLiang) [inline ts/client/features/genesis.ts:52]: Curious about what would happen if we don't add the blockTag?

File: ts/client/features/genesis.ts [hunk 5]
```diff
@@ -17,14 +22,49 @@ export class GenesisCommitment extends BaseCommitment {
...
+        }
+    }
+
+    public async parseBatch(event: Event): Promise<Batch> {
+        const batchIDBN = event.args?.batchID as BigNumber;
+        const batchID = batchIDBN.toNumber();
+
+        this.validateIsFirstBatch(batchID);
+
+        // Create L2 state
```
Comment (ChihChengLiang) [inline ts/client/features/genesis.ts:45]: Nit: What do you think we rephrase this as something like "reconstruct genesis batch from genesis state root"?

File: ts/client/features/genesis.ts [hunk 6]
```diff
@@ -17,14 +22,49 @@ export class GenesisCommitment extends BaseCommitment {
...
         const commitment = new GenesisCommitment(this.genesisStateRoot);
-        return new ConcreteBatch([commitment]);
+        const batch = new ConcreteBatch([commitment]);
+
+        // Get L1 state
```
Comment (ChihChengLiang) [inline ts/client/features/genesis.ts:49]: Nit
```suggestion
        // Get genesis batch from L1
```

**Review by ChihChengLiang — APPROVED** [review]
Looks great! Now the client will validate the genesis batch before doing anything further.
